### ISSUE TITLE
Problem with glob in item transformation.

### ISSUE THREAD

**Issue by mfilippov** [Microsoft/msbuild#3274]
I have problem with glob expanding in item transformation expression. Have you any IDEA how can i fix it?
### Steps to reproduce
Project file
```xml
<?xml version="1.0" encoding="utf-8"?>
<Project xmlns="http://schemas.microsoft.com/developer/msbuild/2003" ToolsVersion="15.0">
  <PropertyGroup>
    <ParserGenOutputBase>C:\Work\test\Core\Psi\Gen\</ParserGenOutputBase>
  </PropertyGroup>
  <ItemGroup>
    <PsiLanguageNames Include="TypeScript;JsDoc;JavaScript;Css">
      <Visible>False</Visible>
      <LanguageName>%(Identity)</LanguageName>
    </PsiLanguageNames>
    <ParserGen Include="Src\JavaScript\Parsing.TypeScript\TypeScript.psi" />
    <ParserGen Include="Src\JavaScript\Parsing.JsDoc\JsDoc.psi" />
    <ParserGen Include="Src\JavaScript\Parsing\JavaScript.psi" />
    <ParserGen Include="Src\Css\Parsing\Css.psi" />

    <Compile Include="@(PsiLanguageNames->'$(ParserGenOutputBase)Parsing\%(Identity)\**\*.cs')">
      <Link>Src\Gen\Css\%(LanguageName)\Psi\Stub</Link>
      <Visible>True</Visible>
    </Compile>
    
  </ItemGroup>
  <Target Name="Build">
    <Message Text="@(Compile)" />
  </Target>
</Project>
```

### Expected  behavior
Glob in "Compile" item group expanded in file list

### Actual behavior
"Compile" item group contains items with TransformedPath\**\*.cs include path.

### Environment data
`msbuild /version` output:
Microsoft (R) Build Engine version 15.6.85.37198 for .NET Framework
Copyright (C) Microsoft Corporation. All rights reserved.

15.6.85.37198
OS info:
Windows

**Comment by rainersigwald** [comment]
Any chance there are files in those directories with long paths, making this an instance of #406?

**Comment by mfilippov** [comment]
@rainersigwald No, I check it. I found the similar issue on StackOverflow: https://stackoverflow.com/questions/48868060/can-a-task-itemgroup-glob-files I think problem connected with the globe in item transformation.

**Comment by mfilippov** [comment]
@rainersigwald I think I could create a test for repro this issue.

**Comment by ViktorHofer** [comment]
Just stumbled over this issue. Having the exact same issue. It seems item transformation treats glob characters as literals.

**Comment by ViktorHofer** [comment]
The trick with the intermediate property from https://stackoverflow.com/questions/48868060/can-a-task-itemgroup-glob-files only work inside a target in case that helps root causing this.

**Comment by dougbu** [comment]
/fyi @rainersigwald we just hit this w/ a transform in dotnet/aspnetcore and are applying the workaround. But, what's the status of this issue❔ It seems to occur with the 6.0.100 SDK.

**Comment by rainersigwald** [comment]
This is happening because MSBuild is escaping the return value of the item transformation, as it does with basically all property-function/transformation returns. Generally you can wrap with the `[MSBuild]::Unescape()` property function, but that doesn't work here because `error MSB4012: The expression "$([MSBuild]::Unescape(@(Folders->'%(Identity)\*.txt'))" cannot be used in this context. Item lists cannot be concatenated with other strings where an item list is expected. Use a semicolon to separate multiple item lists.`

Which is true enough, I suppose. Perhaps we should consider adding an `Unescape` _item function_, though, since there's no way to do this at evaluation time. The first-flatten-it-to-a-property-then-create-an-item-from-that approach is a nice workaround in a target.

**Comment by Turnerj** [comment]
Even when you are in a target though, you can get into difficult scenarios where you can't use that property work around like the OP's example as the item included and its meta property are bound together by both using "LanguageName".

I have a similar issue with my own project where I tried doing this within a target:

```
<PropertyGroup>
	<_BuildVersioningInclude>@(_TargetFrameworks->'../TurnerSoftware.BuildVersioning.Tool/bin/$(Configuration)/%(Identity)/*')</_BuildVersioningInclude>
	<_BuildVersioningPackagePath>@(_TargetFrameworks->'tools/%(Identity)')</_BuildVersioningPackagePath>
</PropertyGroup>
<ItemGroup>
	<None Include="$(_BuildVersioningInclude)" Pack="true" PackagePath="$(_BuildVersioningPackagePath)" />
</ItemGroup>
```

The end result was having each include have multiple package paths where I want one package path per include.
Really what I want to do is this (below) but I can't for the same globbing issue.

```
<None Include="@(_TargetFrameworks->'../TurnerSoftware.BuildVersioning.Tool/bin/$(Configuration)/%(Identity)/*')" Pack="true" PackagePath="tools/%(_TargetFrameworks.Identity)" />
```

**Comment by dougbu** [comment]
Hit this again because I forgot about the "only works inside a target" caveat @ViktorHofer mentioned [above](https://github.com/dotnet/msbuild/issues/3274#issuecomment-897734202). The case was simpler than an item transformation, underscoring @rainersigwald's "basically all property-function/transformation returns" point in [another comment](https://github.com/dotnet/msbuild/issues/3274#issuecomment-976790021).

----

``` msbuild
<PropertyGroup>
    <HelixContentPackageSource>$([MSBuild]::NormalizePath('$(ArtifactsShippingPackagesDir)', '*$(SharedFxVersion).nupkg'))</HelixContentPackageSource>
</PropertyGroup>
<ItemGroup>
    <HelixContent Include="$(HelixContentPackageSource)"/>
</ItemGroup>
```

According to a binary log, `$(HelixContentPackageSource)` has a correct value e.g. `{...}\artifacts\packages\Debug\Shipping\Microsoft.AspNetCore.App.*7.0.0-dev.nupkg` but the `@(HelixContent)` item added _also_ has an Asterix in it.

(I had an easy way out (`$(ArtifactsShippingPackagesDir)*$(SharedFxVersion).nupkg`) in this case because `'$(ArtifactsShippingPackagesDir)` is guaranteed to end with an appropriate slash. Still a gotcha that wasted part of my day.)
